tunnel-obfuscation/src/shadowsocks.rs

```diff
@@ -0,0 +1,258 @@
+use super::Obfuscator;
+use async_trait::async_trait;
+use shadowsocks::{
+    config::{ServerConfig, ServerType},
+    context::Context,
+    crypto::CipherKind,
+    net::ConnectOpts,
+    relay::{udprelay::proxy_socket::ProxySocketError, Address},
+    ProxySocket,
+};
+use std::{io, net::SocketAddr, sync::Arc};
+use tokio::{net::UdpSocket, sync::oneshot};
+
+const SHADOWSOCKS_CIPHER: CipherKind = CipherKind::AES_256_GCM;
+const SHADOWSOCKS_PASSWORD: &str = "mullvad";
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Failed to bind local UDP socket
+    #[error("Failed to bind UDP socket")]
+    BindUdp(#[source] io::Error),
+    /// Missing UDP listener address
+    #[error("Failed to retrieve UDP socket bind address")]
+    GetUdpLocalAddress(#[source] io::Error),
+    /// Failed to wait for UDP client
+    #[error("Failed to wait for UDP client")]
+    WaitForUdpClient(#[source] io::Error),
+    /// Failed to create UDP stream
+    #[error("Failed to create UDP stream")]
+    CreateUdpStream(#[source] io::Error),
+    /// Failed to connect to Shadowsocks endpoint
+    #[error("Failed to connect to Shadowsocks endpoint")]
+    ConnectShadowsocks(#[source] ProxySocketError),
+}
+
+pub struct Shadowsocks {
+    udp_client_addr: SocketAddr,
+    server: tokio::task::JoinHandle<Result<()>>,
+    _shutdown_tx: oneshot::Sender<()>,
```

Comment:
This is unused, I guess you're leaving it in for flexibility in the future?

Reply:
I'm relying on it being dropped for the receiver to resolve (to an abort error). Maybe explicitly sending in `Drop` would be nicer?

Reply:
Yes, that wasn't obvious to me at least. Either a drop impl or a comment to clarify.

Reply:
Added a comment.